lib/SILOptimizer/Mandatory/Differentiation/DerivativeLookup.cpp

```diff
@@ -0,0 +1,122 @@
+//===--- DerivativeLookup.cpp ---------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// SWIFT_ENABLE_TENSORFLOW
+//
+// Utilities for looking up derivatives of functions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "DerivativeLookup.h"
+
+namespace swift {
+
+/// Returns the AbstractFunctionDecl corresponding to `F`. If there isn't one,
+/// returns `nullptr`.
+static AbstractFunctionDecl *getAFDOrNull(SILFunction *F) {
+  auto *DC = F->getDeclContext();
+  if (!DC)
+    return nullptr;
+
+  auto *D = DC->getAsDecl();
+  if (!D)
+    return nullptr;
+
+  return dyn_cast<AbstractFunctionDecl>(D);
+}
+
+SILDifferentiabilityWitness *
```

Comment:
Shouldn't all of these helpers be internal (`static`) symbols, since they are only used by the differentiation transform? I recall that we have a `Differentiation.h` in `Mandatory/`. Can these be added to that file instead?

Reply:
+1 to moving these helpers to `Differentiation.h` under a "Derivative lookup" section.

Reply:
We could move these back for now, because separating files is orthogonal to the purpose of this PR.

I do want to explore separating things into separate compilation units eventually so that:
* incremental compilation is faster
* "private" helper functions (e.g. the getAFD thing in this file) are not visible to all of the differentiation code

Reply:
I think separation is good, but there's probably a better place to put these files. Since `differentiability_witness` and `differentiability_witness_function` are now formal SIL constructs, I believe utilities on these constructs belong to SIL transform utilities (swift/SILOptimizer/Utils/).

Reply:
I moved `DerivativeLookup.{h,cpp}` to `SILOptimizer/Utils` for now.
TF-993 tracks further organization of differentiation code, we can certainly revisit later.